File: matrix-graph/max_area_of_island.js
/**
 * @param {number[][]} grid
 * @return {number}
 */

const explore = (grid, n, m, i, j, result) => {
    if (i<0 || i>=n || j<0 || j>=m) {
        return;
    }
    if (grid[i][j] === 1) {
            grid[i][j] = 0;
            result.areaSoFar += 1;
            result.maxAreaSoFar = Math.max(result.areaSoFar, result.maxAreaSoFar);
            explore(grid, n, m, i-1, j, result);
            explore(grid, n, m, i, j+1, result);
            explore(grid, n, m, i+1, j, result);
            explore(grid, n, m, i, j-1, result);
        
    } else {
        return;
    }
}


var maxAreaOfIsland = function(grid) {
    let n = grid.length;
    let m = grid[0].length;
    
    let result = {
        maxAreaSoFar: 0,
        areaSoFar: 0
    }
    
    for (let i=0; i<n; i++) {
        for (let j=0; j<m; j++) {
            if (grid[i][j] === 1) {
                result.areaSoFar = 0;
                explore(grid, n, m, i, j, result);
            }
        }
    }
    return result.maxAreaSoFar;
    
};

File: matrix-graph/max_area_of_island.py
class Solution:
    def explore(self, grid, n, m, i ,j, result):
        grid[i][j] = 0
        result[0] += 1
        if result[0] > result[1]:
            result[1] = result[0]

        # up
        if (i-1)>=0 and grid[i-1][j] == 1:
            self.explore(grid, n, m, i-1 ,j, result)
        # right
        if (j+1)<=(m-1) and grid[i][j+1] == 1:
            self.explore(grid, n, m, i, j+1, result)
        
        # down
        if (i+1)<=(n-1) and grid[i+1][j] == 1:
            self.explore(grid, n, m, i+1 ,j, result)
        
        # left
        if j-1>=0 and grid[i][j-1] == 1:
            self.explore(grid, n, m, i, j-1, result)

        
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        if not grid:
            return 0
        n = len(grid)
        m = len(grid[0])
        
        # size_so_far and max_size
        result = [0, 0]
        for i in range(n):
            for j in range(m):
                if grid[i][j] == 1:
                    result[0] = 0
                    self.explore(grid, n, m, i, j, result)
        return result[1]